ADVANCED REAL WORLD DATA ANALYSIS

In [2]:
import pandas as pd

C:\Users\VAIBHAV\AppData\Local\Temp\ipykernel_12632\4080736814.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [3]:
df = pd.read_csv("Data_Co_Supply_Chain_Dataset.csv" , encoding="latin1")
df.shape

(180519, 31)

In [4]:
df.columns

Index(['Benefit per order', 'Sales per customer', 'Delivery Status',
       'Late_delivery_risk', 'Category Name', 'Customer City',
       'Customer Country', 'Customer Fname', 'Customer Id', 'Customer Lname',
       'Customer Segment', 'Customer State', 'Department Name', 'Market',
       'Order Country', 'order date (DateOrders)', 'Order Id',
       'Order Item Discount', 'Order Item Discount Rate',
       'Order Item Product Price', 'Order Item Profit Ratio',
       'Order Item Quantity', 'Sales', 'Order Item Total',
       'Order Profit Per Order', 'Order Region', 'Order State', 'Order Status',
       'Product Name', 'shipping date (DateOrders)', 'Shipping Mode'],
      dtype='object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 31 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Benefit per order           180519 non-null  float64
 1   Sales per customer          180519 non-null  float64
 2   Delivery Status             180519 non-null  object 
 3   Late_delivery_risk          180519 non-null  int64  
 4   Category Name               180519 non-null  object 
 5   Customer City               180519 non-null  object 
 6   Customer Country            180519 non-null  object 
 7   Customer Fname              180519 non-null  object 
 8   Customer Id                 180519 non-null  int64  
 9   Customer Lname              180511 non-null  object 
 10  Customer Segment            180519 non-null  object 
 11  Customer State              180519 non-null  object 
 12  Department Name             180519 non-null  object 
 13  Market        

In [6]:
df.describe()

,Benefit per order,Sales per customer,Late_delivery_risk,Customer Id,Order Id,Order Item Discount,Order Item Discount Rate,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order
count,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000
mean,21.974989,183.107609,0.548291,6691.379495,36221.894903,20.664741,0.101668,141.232550,0.120647,2.127638,203.772096,183.107609,21.974989
std,104.433526,120.043670,0.497664,4162.918106,21045.379569,21.800901,0.070415,139.732492,0.466796,1.453451,132.273077,120.043670,104.433526
min,-4274.979980,7.490000,0.000000,1.000000,1.000000,0.000000,0.000000,9.990000,-2.750000,1.000000,9.990000,7.490000,-4274.979980
25%,7.000000,104.379997,0.000000,3258.500000,18057.000000,5.400000,0.040000,50.000000,0.080000,1.000000,119.980003,104.379997,7.000000
50%,31.520000,163.990005,1.000000,6457.000000,36140.000000,14.000000,0.100000,59.990002,0.270000,1.000000,199.919998,163.990005,31.520000
75%,64.800003,247.399994,1.000000,9779.000000,54144.000000,29.990000,0.160000,199.990005,0.360000,3.000000,299.950012,247.399994,64.800003
max,911.799988,1939.989990,1.000000,20757.000000,77204.000000,500.000000,0.250000,1999.989990,0.500000,5.000000,1999.989990,1939.989990,911.799988


DATA QUALITY ANALYSIS

In [7]:
df.isnull().sum()

Benefit per order             0
Sales per customer            0
Delivery Status               0
Late_delivery_risk            0
Category Name                 0
Customer City                 0
Customer Country              0
Customer Fname                0
Customer Id                   0
Customer Lname                8
Customer Segment              0
Customer State                0
Department Name               0
Market                        0
Order Country                 0
order date (DateOrders)       0
Order Id                      0
Order Item Discount           0
Order Item Discount Rate      0
Order Item Product Price      0
Order Item Profit Ratio       0
Order Item Quantity           0
Sales                         0
Order Item Total              0
Order Profit Per Order        0
Order Region                  0
Order State                   0
Order Status                  0
Product Name                  0
shipping date (DateOrders)    0
Shipping Mode                 0
dtype: i

In [8]:
df.duplicated().sum()

0

In [9]:
df["Delivery Status"].value_counts()

Delivery Status
Late delivery        98977
Advance shipping     41592
Shipping on time     32196
Shipping canceled     7754
Name: count, dtype: int64

In [10]:
df["Late_delivery_risk"].value_counts()

Late_delivery_risk
1    98977
0    81542
Name: count, dtype: int64

In [11]:
df["Order Status"].value_counts()

Order Status
COMPLETE           59491
PENDING_PAYMENT    39832
PROCESSING         21902
PENDING            20227
CLOSED             19616
ON_HOLD             9804
SUSPECTED_FRAUD     4062
CANCELED            3692
PAYMENT_REVIEW      1893
Name: count, dtype: int64

In [14]:
df[["order date (DateOrders)" ,"shipping date (DateOrders)"]].head()

,order date (DateOrders),shipping date (DateOrders)
0,1/31/2018 22:56,02-03-2018 22:56
1,1/13/2018 12:27,1/18/2018 12:27
2,1/13/2018 12:06,1/17/2018 12:06
3,1/13/2018 11:45,1/16/2018 11:45
4,1/13/2018 11:24,1/15/2018 11:24


In [20]:
df["order date (DateOrders)"] = pd.to_datetime(
    df["order date (DateOrders)"] ,
    format="mixed"
)

df["shipping date (DateOrders)"] = pd.to_datetime(
    df["shipping date (DateOrders)"],
    format="mixed"
)

In [21]:
df[["order date (DateOrders)" ,"shipping date (DateOrders)"]].head()

,order date (DateOrders),shipping date (DateOrders)
0,2018-01-31 22:56:00,2018-02-03 22:56:00
1,2018-01-13 12:27:00,2018-01-18 12:27:00
2,2018-01-13 12:06:00,2018-01-17 12:06:00
3,2018-01-13 11:45:00,2018-01-16 11:45:00
4,2018-01-13 11:24:00,2018-01-15 11:24:00


In [23]:
# Date features
df["Order Year"] = df["order date (DateOrders)"].dt.year
df["Order Month"] = df["order date (DateOrders)"].dt.month
df["Order Month Name"] = df["order date (DateOrders)"].dt.month_name()
df["Order Day Name"] = df["order date (DateOrders)"].dt.day_name()

In [26]:
df[[
      "order date (DateOrders)",
          "Order Year",
          "Order Month",
          "Order Month Name",
          "Order Day Name"
]].head()

,order date (DateOrders),Order Year,Order Month,Order Month Name,Order Day Name
0,2018-01-31 22:56:00,2018,1,January,Wednesday
1,2018-01-13 12:27:00,2018,1,January,Saturday
2,2018-01-13 12:06:00,2018,1,January,Saturday
3,2018-01-13 11:45:00,2018,1,January,Saturday
4,2018-01-13 11:24:00,2018,1,January,Saturday


REGIONAL SEGEMNT PERFORMANCE

In [28]:
region_analysis = df.groupby("Order Region").agg(
     Total_Sales = ("Sales" , "sum"),
     Total_Profit= ("Order Profit Per Order" , "sum"),
     Average_Sales= ("Sales" , "mean"),
     Average_Profit= ("Order Profit Per Order" , "mean"),
     Total_Orders = ("Order Id" , "nunique")
)
region_analysis.sort_values(
     by="Total_Orders",
     ascending=True
)

,Total_Sales,Total_Profit,Average_Sales,Average_Profit,Total_Orders
Order Region,,,,,
Central Asia,1.098399e+05,13045.280051,198.625556,23.590018,184
Canada,1.868610e+05,23900.709960,194.849889,24.922534,309
Southern Africa,2.282516e+05,30826.050146,197.278820,26.643086,398
Central Africa,3.272630e+05,33447.269960,195.147893,19.944705,556
East Africa,3.762349e+05,43167.729927,203.150593,23.308709,613
North Africa,6.347522e+05,64599.860009,196.396108,19.987580,1064
West Africa,7.279512e+05,80030.270031,196.956492,21.653212,1223
Eastern Europe,7.742666e+05,79717.049920,197.516981,20.335982,1292
South of USA,7.857839e+05,88114.879898,194.260555,21.783654,1345


Western Europe has the highest total sales: ~5.89M

Western Europe also has the highest total profit: ~625.45K

Western Europe has the highest average sales: ~217.43

Central America has 9,396 orders, making it another major region by order volume.

Northern Europe has the highest average profit among the displayed regions: ~23.84

In [29]:
# identify the best region 
region_analysis["Total_Orders"].idxmax()

'Western Europe'

In [31]:
region_analysis["Total_Sales"].idxmax()

'Western Europe'

In [32]:
region_analysis["Total_Profit"].idxmax()

'Western Europe'

CUSTOMER SEGMENT

In [33]:
segment_analysis = df.groupby("Customer Segment").agg(
    Total_Sales=("Sales", "sum"),
    Total_Profit=("Order Profit Per Order", "sum"),
    Average_Sales=("Sales", "mean"),
    Average_Profit=("Order Profit Per Order", "mean"),
    Total_Orders=("Order Id", "nunique")
)
segment_analysis.sort_values(
    by="Total_Sales",
    ascending=False
)

,Total_Sales,Total_Profit,Average_Sales,Average_Profit,Total_Orders
Customer Segment,,,,,
Consumer,1.909579e+07,2.073488e+06,204.224313,22.175390,34119
Corporate,1.116841e+07,1.202575e+06,203.843962,21.949204,19856
Home Office,6.520538e+06,6.908403e+05,202.337802,21.437359,11777


In [34]:
segment_analysis["Total_Sales"].idxmax()

'Consumer'

In [35]:
segment_analysis["Total_Profit"].idxmax()

'Consumer'

In [36]:
segment_analysis["Total_Orders"].idxmax()

'Consumer'

SALES CONTRIBUTION BY SEGMENT

In [37]:
total_sales = segment_analysis["Total_Sales"].sum()
total_sales

36784735.01337984

In [42]:
segment_analysis["Sales_Contribution_%"] = (
     segment_analysis["Total_Sales"]
     .div(total_sales)
     .mul(100)
)
segment_analysis["Sales_Contribution_%"]

Customer Segment
Consumer       51.912268
Corporate      30.361526
Home Office    17.726206
Name: Sales_Contribution_%, dtype: float64

PROFIT CONTRIBUTION BY SEGMENT

In [43]:
segment_analysis["Profit_Contribution_%"] = (
    segment_analysis["Total_Profit"]
    .div(segment_analysis["Total_Profit"].sum())
    .mul(100)
)

In [44]:
segment_analysis[
    ["Total_Sales", "Sales_Contribution_%", "Total_Profit", "Profit_Contribution_%"]
].sort_values(
    by="Profit_Contribution_%",
    ascending=False
)

,Total_Sales,Sales_Contribution_%,Total_Profit,Profit_Contribution_%
Customer Segment,,,,
Consumer,1.909579e+07,51.912268,2.073488e+06,52.269685
Corporate,1.116841e+07,30.361526,1.202575e+06,30.315210
Home Office,6.520538e+06,17.726206,6.908403e+05,17.415106


Category Performance

In [45]:
category_analysis = df.groupby("Category Name").agg(
    Total_Sales=("Sales", "sum"),
    Total_Profit=("Order Profit Per Order", "sum"),
    Average_Sales=("Sales", "mean"),
    Average_Profit=("Order Profit Per Order", "mean"),
    Total_Orders=("Order Id", "nunique")
)
category_analysis.sort_values(
    by="Total_Sales",
    ascending=False
).head(10)

,Total_Sales,Total_Profit,Average_Sales,Average_Profit,Total_Orders
Category Name,,,,,
Fishing,6.929654e+06,756220.767190,399.980011,43.649106,15164
Cleats,4.431943e+06,494636.919791,180.519848,20.147323,20386
Camping & Hiking,4.118426e+06,427455.568106,299.980011,31.135230,12299
Cardio Equipment,3.694843e+06,383011.098485,295.895187,30.672788,11355
Women's Apparel,3.147800e+06,350421.029567,149.645828,16.658951,17869
Water Sports,3.113845e+06,325146.960038,200.376106,20.923228,13758
Men's Footwear,2.891758e+06,311902.820214,129.990005,14.020625,18783
Indoor/Outdoor Games,2.888994e+06,318451.430554,149.704317,16.501784,16623
Shop By Sport,1.309522e+06,129813.960315,119.220870,11.818460,10136


In [46]:
category_analysis["Total_Sales"].idxmax()

'Fishing'

In [47]:
category_analysis["Total_Profit"].idxmax()

'Fishing'

In [48]:
category_analysis["Average_Profit"].idxmax()

'Computers'

In [49]:
category_analysis["Total_Orders"].idxmax()

'Cleats'

Sales vs Profit Analysis

In [50]:
category_analysis["Profit_Margin_%"] = (
    category_analysis["Total_Profit"]
    .div(category_analysis["Total_Sales"])
    .mul(100)
)
category_analysis.sort_values(
    by="Profit_Margin_%",
    ascending=False
)

,Total_Sales,Total_Profit,Average_Sales,Average_Profit,Total_Orders,Profit_Margin_%
Category Name,,,,,,
Golf Bags & Carts,1.036939e+04,1810.069992,169.990005,29.673279,61,17.455896
Fitness Accessories,3.560144e+04,5258.390016,115.215018,17.017443,307,14.770161
Toys,6.104660e+03,900.710000,11.540000,1.702665,529,14.754466
Soccer,2.647705e+04,3901.949953,191.862680,28.275000,136,14.737102
Women's Clothing,1.402830e+05,19102.849930,215.820007,29.389000,650,13.617366
Baseball & Softball,9.405715e+04,12762.130067,148.824608,20.193244,628,13.568484
Garden,2.577687e+05,33443.010106,532.580017,69.097128,484,12.974037
Tennis & Racquet,4.458509e+04,5747.979996,135.930154,17.524329,327,12.892157
Music,1.131221e+05,14436.319923,260.649994,33.263410,434,12.761715


In [51]:
category_analysis["Profit_Margin_%"].idxmax()

'Golf Bags & Carts'

In [52]:
category_analysis["Profit_Margin_%"].idxmin()

'Strength Training'

Delivery Risk Analysis

In [53]:
delivery_risk_region = df.groupby("Order Region")[
    "Late_delivery_risk"
].mean().mul(100)
delivery_risk_region.sort_values(
    ascending=False
)

Order Region
Central Africa     57.960644
South Asia         56.266977
East Africa        55.939525
Western Europe     55.848611
South of  USA      55.772559
Eastern Europe     55.663265
East of USA        55.661605
Southeast Asia     55.529930
Central Asia       55.334539
West Asia          55.283741
US Center          55.240360
Central America    54.754596
North Africa       54.517327
Southern Europe    54.384477
Eastern Asia       54.326923
South America      54.308671
Northern Europe    54.044118
Oceania            54.020497
West of USA        53.959715
Southern Africa    53.327571
Caribbean          53.077663
West Africa        52.840909
Canada             48.800834
Name: Late_delivery_risk, dtype: float64

Within-Region Segment Contribution

In [56]:
region_segment = df.groupby(
    ["Order Region", "Customer Segment"]
)["Sales"].sum().reset_index()

In [57]:
region_segment["Region_Total_Sales"] = (
    region_segment.groupby("Order Region")["Sales"]
    .transform("sum")
)

In [58]:
region_segment["Sales_Contribution_%"] = (
    region_segment["Sales"]
    .div(region_segment["Region_Total_Sales"])
    .mul(100)
)

In [59]:
dominant_segment = region_segment.loc[
    region_segment.groupby("Order Region")["Sales_Contribution_%"].idxmax()
]

In [61]:
dominant_segment.sort_values(
    by="Sales_Contribution_%",
    ascending=False
)

,Order Region,Customer Segment,Sales,Region_Total_Sales,Sales_Contribution_%
12,Central Asia,Consumer,6.457016e+04,1.098399e+05,58.785689
54,US Center,Consumer,6.233922e+05,1.151356e+06,54.144183
15,East Africa,Consumer,2.030225e+05,3.762349e+05,53.961628
42,South of USA,Consumer,4.213899e+05,7.857839e+05,53.626685
24,Eastern Europe,Consumer,4.105639e+05,7.742666e+05,53.026164
27,North Africa,Consumer,3.336980e+05,6.347522e+05,52.571381
60,West Asia,Consumer,6.154214e+05,1.174672e+06,52.390920
66,Western Europe,Consumer,3.084309e+06,5.894381e+06,52.326253
33,Oceania,Consumer,1.051469e+06,2.016654e+06,52.139295
9,Central America,Consumer,2.953943e+06,5.665712e+06,52.137182


## Final Business Insights — Day 21

1. Western Europe is the strongest-performing region, leading in total sales, total profit, and total orders.

2. Consumer is the strongest customer segment, contributing approximately 51.91% of total sales and 52.27% of total profit.

3. Fishing is the highest-performing category by total sales and total profit, while Computers has the highest average profit per order.

4. Golf Bags & Carts has the highest profit margin at approximately 17.46%, while Strength Training has the lowest at approximately 0.61%.

5. Central Africa has the highest late-delivery risk at approximately 57.96%, indicating a potential supply-chain performance concern.

6. Consumer is the dominant sales segment across most regions, but its contribution varies by region. Central Africa, for example, has a Consumer contribution of approximately 49.17%.